File: app/src/main/java/me/tysheng/xishi/adapter/BaseDataBindingAdapter.java
package me.tysheng.xishi.adapter;

import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.support.annotation.LayoutRes;
import android.support.annotation.Nullable;
import android.view.View;
import android.view.ViewGroup;

import com.chad.library.adapter.base.BaseQuickAdapter;

import java.util.List;

/**
 * Created by tysheng
 * Date: 2017/5/11 14:39.
 * Email: tyshengsx@gmail.com
 */

public abstract class BaseDataBindingAdapter<T, Binding extends ViewDataBinding> extends BaseQuickAdapter<T, TyBindingViewHolder<Binding>> {


    public BaseDataBindingAdapter(@LayoutRes int layoutResId, @Nullable List<T> data) {
        super(layoutResId, data);
    }

    public BaseDataBindingAdapter(@Nullable List<T> data) {
        super(data);
    }

    public BaseDataBindingAdapter(@LayoutRes int layoutResId) {
        super(layoutResId);
    }

    @Override
    protected TyBindingViewHolder<Binding> createBaseViewHolder(View view) {
        return new TyBindingViewHolder<>(view);
    }

    @Override
    protected TyBindingViewHolder<Binding> createBaseViewHolder(ViewGroup parent, int layoutResId) {
        Binding binding = DataBindingUtil.inflate(mLayoutInflater, layoutResId, parent, false);
        View view;
        if (binding == null) {
            view = getItemView(layoutResId, parent);
        } else {
            view = binding.getRoot();
        }
        TyBindingViewHolder<Binding> holder = new TyBindingViewHolder<>(view);
        holder.setBinding(binding);
        return holder;
    }

    @Override
    protected void convert(TyBindingViewHolder<Binding> helper, T item) {
        convert(helper.getBinding(), item);
        helper.getBinding().executePendingBindings();
    }

    public void onEnd() {
        loadMoreEnd();
    }

    public void onError() {
        loadMoreFail();
    }

    protected abstract void convert(Binding binding, T item);
}

File: app/src/main/java/me/tysheng/xishi/adapter/BaseDataBindingAdapter.kt
package me.tysheng.xishi.adapter

import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.support.annotation.LayoutRes
import android.view.View
import android.view.ViewGroup

import com.chad.library.adapter.base.BaseQuickAdapter

/**
 * Created by tysheng
 * Date: 2017/5/11 14:39.
 * Email: tyshengsx@gmail.com
 */

abstract class BaseDataBindingAdapter<T, Binding : ViewDataBinding> : BaseQuickAdapter<T, TyBindingViewHolder<Binding>> {


    constructor(@LayoutRes layoutResId: Int, data: List<T>?) : super(layoutResId, data) {}

    constructor(data: List<T>?) : super(data) {}

    constructor(@LayoutRes layoutResId: Int) : super(layoutResId) {}

    override fun createBaseViewHolder(view: View): TyBindingViewHolder<Binding> {
        return TyBindingViewHolder(view)
    }

    override fun createBaseViewHolder(parent: ViewGroup, layoutResId: Int): TyBindingViewHolder<Binding> {
        val binding = DataBindingUtil.inflate<Binding>(mLayoutInflater, layoutResId, parent, false)
        val view: View
        if (binding == null) {
            view = getItemView(layoutResId, parent)
        } else {
            view = binding.root
        }
        val holder = TyBindingViewHolder<Binding>(view)
        holder.binding = binding
        return holder
    }

    override fun convert(helper: TyBindingViewHolder<Binding>, item: T) {
        convert(helper.binding, item)
        helper.binding!!.executePendingBindings()
    }

    fun onEnd() {
        loadMoreEnd()
    }

    fun onError() {
        loadMoreFail()
    }

    protected abstract fun convert(binding: Binding?, item: T)
}
